### PR TITLE
feat: reference implementation for Hydra+NextJS

### PR DESCRIPTION
**Pull request by amorey**
## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md) and signed the CLA.
- [ ] I have referenced an issue containing the design document if my change introduces a new feature.
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security vulnerability. If this pull request addresses a security
      vulnerability, I confirm that I got green light (please contact [security@ory.sh](mailto:security@ory.sh)) from the
      maintainers to push the changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation within the code base (if appropriate).

### CODE REVIEW

**Comment by vinckr** [comment]
Thanks so much for these examples @amorey !
I will see to review next week latest 🙌 
If you like you can also take a look @kevgo

File: hydra-nextjs/LICENSE.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Copyright 2022 Andres Morey
```
Comment (kevgo) [inline hydra-nextjs/LICENSE.txt:1]: This repo already has a [license in the top-level folder](https://github.com/ory/examples/blob/master/LICENSE) (Apache 2) and the CLA transfers the copyright to Ory. Can you please remove this license file?

File: hydra-nextjs/middleware.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import csrf from 'edge-csrf';
+import { NextResponse } from 'next/server';
```
Comment (kevgo) [inline hydra-nextjs/middleware.js:2]: `.js` files cannot use `import` statements. Rename to `middleware.mjs`?

File: hydra-nextjs/README.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+This application can be configured using the following environment variables:
+
+| Name                    | Default                |
+| ----------------------- | ---------------------- |
+| SECURITY_HEADERS_ENABLE | false                  |
+| HYDRA_ADMIN_URL         | http://localhost:4445/ |
+
+## Development
+
+To install dependencies:
```
Comment (kevgo) [inline hydra-nextjs/README.md:27]: You you please describe how to run this using the Dockerfile it this PR?

File: hydra-nextjs/cypress/e2e/headers.cy.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      'permissions-policy': 'accelerometer=(),autoplay=(),camera=(),display-capture=(),document-domain=(),encrypted-media=(),fullscreen=(),geolocation=(),gyroscope=(),magnetometer=(),microphone=(),midi=(),payment=(),picture-in-picture=(),publickey-credentials-get=(),screen-wake-lock=(),sync-xhr=(self),usb=(),web-share=(),xr-spatial-tracking=()'
+    };
+
+    const response = await cy.request('GET', '/');
+    for (const [key, val] of Object.entries(securityHeaders)) {
+      cy.log(key);
+      expect(response.headers).to.have.property(key, val);
+    }
+  });
+});
```
Comment (kevgo) [inline hydra-nextjs/cypress/e2e/headers.cy.js:27]: Can you please add an end-to-end test for the happy path, i.e. one that successfully signs up and logs in?

File: hydra-nextjs/tests/auth/#sign-in.test.jsx# [hunk 5]
```diff
@@ -0,0 +1,344 @@
+Aimport '@testing-library/jest-dom';
```
Comment (kevgo) [inline hydra-nextjs/tests/auth/#sign-in.test.jsx#]: typo?

```suggestion
import '@testing-library/jest-dom';
```

File: hydra-nextjs/tests/helpers.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+export function mockRequest(options) {
+  let req = new http.IncomingMessage();
+
+  const { body, ...attrs } = options || {};
+  req = Object.assign(req, attrs);
+
+  // write body
+  req.push(body || '');
+  req.push(null);
```
Comment (kevgo) [inline hydra-nextjs/tests/helpers.js:11]: Does pushing `null` do something useful here?

File: hydra-nextjs/README.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+```sh
+yarn install
+```
+
+To run the Next.js app server in development mode:
+
+```sh
+yarn dev
```
Comment (kevgo) [inline hydra-nextjs/README.md:36]: When I run `yarn install` and `yarn dev` on my system (Debian), I see `localhost:3000` but it always displays "Login challenge is missing" and neither the sign-up nor sign-in page work. Am I missing some setup? If yes, can you please document it here?

**Review by kevgo — COMMENTED** [review]
Thanks for this great example and the comprehensive test coverage! How do you feel about adding an end-to-end test for the happy path, i.e. a Cypress test that boots up the entire stack (no mocks) and demonstrates how to sign up, log out, and sign in again? I think that would add a lot of value that this app works and allow refactoring with confidence.

**Comment by kevgo** [comment]
To satisfy the `Conventional commits` check, your commit message body must start with a lowercase character. Renaming this PR to `feat: reference implementation for Hydra+NextJS` should do it.

File: hydra-nextjs/tests/helpers.js (hunk 6 above)
Comment (amorey) [reply hydra-nextjs/tests/helpers.js:11]: Pushing `null` ends the `IncomingMessage`, however I think I'll switch to the node-mocks-http library which will replace this custom code

File: hydra-nextjs/cypress/e2e/headers.cy.js (hunk 4 above)
Comment (amorey) [reply hydra-nextjs/cypress/e2e/headers.cy.js:27]: The happy path requires a running hydra server and an oauth client. Do you have any recommendations for how to do this in an e2e test?

File: hydra-nextjs/README.md (hunk 7 above)
Comment (amorey) [reply hydra-nextjs/README.md:36]: You need an oauth client to initiate the authentication flow and a running hydra server to provide a valid challenge. How would you recommend providing an oauth client for this project? Is there a docker image that I could use?

**Review by vinckr — APPROVED** [review]
we can improve on this with tests & optimized docker setup later on, it seems to work fine with Hydra running

Thanks so much for the contribution and apologies for the late review @amorey

**Pull request merged**